pkg/check/datastore.go

```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
 }
 
-func checkDataStore(dsName string, infrastructure *configv1.Infrastructure) error {
+func checkDataStore(ctx *CheckContext, dsName string, infrastructure *configv1.Infrastructure) error {
 	clusterID := infrastructure.Status.InfrastructureName
 	volumeName := fmt.Sprintf("[%s] 00000000-0000-0000-0000-000000000000/%s-dynamic-pvc-00000000-0000-0000-0000-000000000000.vmdk", dsName, clusterID)
 	klog.V(4).Infof("Checking data store %q with potential volume Name %s", dsName, volumeName)
 	if err := checkVolumeName(volumeName); err != nil {
 		return fmt.Errorf("datastore %s: %s", dsName, err)
 	}
+	if err := checkForDatastoreCluster(ctx, dsName); err != nil {
+		return err
+	}
 	return nil
 }
 
+func checkForDatastoreCluster(ctx *CheckContext, dataStoreName string) error {
+	tctx, cancel := context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	finder := find.NewFinder(ctx.VMClient, false)
+	datacenters, err := finder.DatacenterList(tctx, "*")
+	if err != nil {
+		klog.Errorf("error listing datacenters: %v", err)
+		return nil
+	}
+	workspaceDC := ctx.VMConfig.Workspace.Datacenter
+	var matchingDC *object.Datacenter
+	for _, dc := range datacenters {
+		if dc.Name() == workspaceDC {
+			matchingDC = dc
+		}
+	}
+
+	// lets fetch the datastore
+	finder = find.NewFinder(ctx.VMClient, false)
+	finder.SetDatacenter(matchingDC)
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	ds, err := finder.Datastore(tctx, dataStoreName)
+	if err != nil {
+		klog.Errorf("error getting datastore %s: %v", dataStoreName, err)
+		return nil
+	}
+
+	var dsMo mo.Datastore
+	pc := property.DefaultCollector(matchingDC.Client())
+	properties := []string{DatastoreInfoProperty, SummaryProperty}
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	err = pc.RetrieveOne(tctx, ds.Reference(), properties, &dsMo)
+
+	if err != nil {
+		klog.Errorf("error getting properties of datastore %s: %v", dataStoreName, err)
+		return nil
+	}
+
+	// list datastore cluster
+	m := view.NewManager(ctx.VMClient)
+	kind := []string{"StoragePod"}
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	v, err := m.CreateContainerView(tctx, ctx.VMClient.ServiceContent.RootFolder, kind, true)
+	if err != nil {
+		klog.Errorf("error listing datastore cluster: %+v", err)
+		return nil
+	}
+	var content []mo.StoragePod
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	err = v.Retrieve(tctx, kind, []string{SummaryProperty, "childEntity"}, &content)
+	if err != nil {
+		klog.Errorf("error retrieving datastore cluster properties: %+v", err)
+		return nil
+	}
+	err = v.Destroy(tctx)
+	if err != nil {
+		klog.Errorf("error destroying view: %+v", err)
+		return nil
+	}
+	for _, ds := range content {
+		for _, child := range ds.Folder.ChildEntity {
+			tDS, err := getDatastore(ctx, child)
+			if err != nil {
+				klog.Errorf("fetching datastore %s failed: %v", child.String(), err)
```

Comment:
Why is the error ignored? A comment would be nice.

Reply:
I could add a comment. It is being ignored because - it is possible that the OCP user may not have permission to fetch datastore that is not being used in OCP.

Reply:
added